syntax = "proto3";

package com.webank.eggroll.core.meta;

message Endpoint {
  string host = 1;
  int32 port = 2;
}

message ServerNode {
  string id = 1;
  Endpoint commandEndpoint = 2;
  Endpoint dataEndpoint = 3;
  string tag = 4;
}

message ServerCluster {
  string id = 1;
  repeated ServerNode nodes = 2;
  string tag = 3;
  bytes partitioner = 4;
}

message Functor {
  string name = 1;
  string serdes = 2;
  bytes body = 3;
  map<string, string> conf = 4;
}

message Pair {
  bytes key = 1;
  bytes value = 2;
}

message PairBatch {
  repeated Pair pairs = 1;
}

message StoreLocator {
  string storeType = 1;
  string namespace = 2;
  string name = 3;
  string path = 4;
}

message Store {
  StoreLocator storeLocator = 1;
  repeated Partition partitions = 2;
}

message Partition {
  string id = 1;
  StoreLocator storeLocator = 2;
  ServerNode node = 3;
}

message CallInfo {
  string callSeq = 1;
}

message Job {
  string id = 1;
  string name = 2;
  repeated Store inputs = 3;
  repeated Store outputs = 4;
  repeated Functor functors = 5;
}

message Task {
  string id = 1;
  string name = 2;
  repeated Partition inputs = 3;
  repeated Partition outputs = 4;
  Job job = 5;
}